Better Input Extension

SharkPool-SP opened the pull request:
**VERSION 4.0!!** 
Adds various blocks for custom CSS inputs. Focused on remodeling the ask () block to be better.
![image](https://github.com/TurboWarp/extensions/assets/139097378/26f54578-14a0-4832-be02-a4dce00854fb)

<img width="1103" alt="Screenshot 2023-10-09 at 9 04 42 PM" src="https://github.com/TurboWarp/extensions/assets/139097378/bf6b1fbe-3c1a-4875-b597-7413e9f8eb63">

SharkPool-SP commented:
Here is the thumbnail that I forgot to include: 
![Better Input](https://github.com/TurboWarp/extensions/assets/139097378/c4a74de2-08a5-4ca2-a2b1-e434031dd0d9)

qxsck commented:
> Adds various blocks for custom CSS inputs. Focused on remodeling the ask () block to be better. <img alt="Screen Shot 2023-07-30 at 11 49 39 PM" width="521" src="https://usercontent.githubfast.com/user-images/139097378/257139207-b57871ee-0814-4817-a927-a5e4d2252018.png">

Should UserInput return null instead of 'canceled' without input?

SharkPool-SP commented:
> > Adds various blocks for custom CSS inputs. Focused on remodeling the ask () block to be better. <img alt="Screen Shot 2023-07-30 at 11 49 39 PM" width="521" src="https://camo.githubusercontent.com/a479a9b812fd016825b9651cfb1690a144b2713d2af251e603950ce9e0a4bcc1/68747470733a2f2f75736572636f6e74656e742e676974687562666173742e636f6d2f757365722d696d616765732f3133393039373337382f3235373133393230372d62353738373165652d303831342d343831372d613932372d6135653464323235323031382e706e67">
> 
> Should UserInput return null instead of 'canceled' without input?

Doesn’t really matter. If there’s no response it will return empty; just like the current ask and wait block

LilyMakesThings commented:
> > > Adds various blocks for custom CSS inputs. Focused on remodeling the ask () block to be better. <img alt="Screen Shot 2023-07-30 at 11 49 39 PM" width="521" src="https://camo.githubusercontent.com/a479a9b812fd016825b9651cfb1690a144b2713d2af251e603950ce9e0a4bcc1/68747470733a2f2f75736572636f6e74656e742e676974687562666173742e636f6d2f757365722d696d616765732f3133393039373337382f3235373133393230372d62353738373165652d303831342d343831372d613932372d6135653464323235323031382e706e67">
> > 
> > 
> > Should UserInput return null instead of 'canceled' without input?
> 
> Doesn’t really matter. If there’s no response it will return empty; just like the current ask and wait block

It does matter, actually. For parity with the actual ask block, you should return an empty string.

SharkPool-SP commented:
> > > > Adds various blocks for custom CSS inputs. Focused on remodeling the ask () block to be better. <img alt="Screen Shot 2023-07-30 at 11 49 39 PM" width="521" src="https://camo.githubusercontent.com/a479a9b812fd016825b9651cfb1690a144b2713d2af251e603950ce9e0a4bcc1/68747470733a2f2f75736572636f6e74656e742e676974687562666173742e636f6d2f757365722d696d616765732f3133393039373337382f3235373133393230372d62353738373165652d303831342d343831372d613932372d6135653464323235323031382e706e67">
> > > 
> > > 
> > > Should UserInput return null instead of 'canceled' without input?
> > 
> > 
> > Doesn’t really matter. If there’s no response it will return empty; just like the current ask and wait block
> 
> It does matter, actually. For parity with the actual ask block, you should return an empty string.

No that’s what I mean, it will return an empty string when there’s no input

LilyMakesThings commented:
> > > > > Adds various blocks for custom CSS inputs. Focused on remodeling the ask () block to be better. <img alt="Screen Shot 2023-07-30 at 11 49 39 PM" width="521" src="https://camo.githubusercontent.com/a479a9b812fd016825b9651cfb1690a144b2713d2af251e603950ce9e0a4bcc1/68747470733a2f2f75736572636f6e74656e742e676974687562666173742e636f6d2f757365722d696d616765732f3133393039373337382f3235373133393230372d62353738373165652d303831342d343831372d613932372d6135653464323235323031382e706e67">
> > > > 
> > > > 
> > > > Should UserInput return null instead of 'canceled' without input?
> > > 
> > > 
> > > Doesn’t really matter. If there’s no response it will return empty; just like the current ask and wait block
> > 
> > 
> > It does matter, actually. For parity with the actual ask block, you should return an empty string.
> 
> No that’s what I mean, it will return an empty string when there’s no input

It should also do that when it's cancelled. Make a boolean that is toggled when an input is cancelled or not.

SharkPool-SP commented:
> > > > > > Adds various blocks for custom CSS inputs. Focused on remodeling the ask () block to be better. <img alt="Screen Shot 2023-07-30 at 11 49 39 PM" width="521" src="https://camo.githubusercontent.com/a479a9b812fd016825b9651cfb1690a144b2713d2af251e603950ce9e0a4bcc1/68747470733a2f2f75736572636f6e74656e742e676974687562666173742e636f6d2f757365722d696d616765732f3133393039373337382f3235373133393230372d62353738373165652d303831342d343831372d613932372d6135653464323235323031382e706e67">
> > > > > 
> > > > > 
> > > > > Should UserInput return null instead of 'canceled' without input?
> > > > 
> > > > 
> > > > Doesn’t really matter. If there’s no response it will return empty; just like the current ask and wait block
> > > 
> > > 
> > > It does matter, actually. For parity with the actual ask block, you should return an empty string.
> > 
> > 
> > No that’s what I mean, it will return an empty string when there’s no input
> 
> It should also do that when it's cancelled. Make a boolean that is toggled when an input is cancelled or not.

I might add the Boolean for that in a later update. I like the cancel button the way it is thought since the button makes the input report cancelled. That way people can use better input as a button selection thing rather than an input and press OK. That’s also why the Input and cancel button has an option to be enabled or disabled.

SharkPool-SP commented:
Extension was updated. Merged some blocks, changed some names and functions, added the ability to enable/disable 2 more buttons and fixed some bugs.

SaFireGalaxy commented:
Hey sharkpool? Can you possibly make it so you can force-submit the input? This would be handy for having the "enter" key be another option to submit your inputs.

SurvExe1Pc commented:
> Hey sharkpool? Can you possibly make it so you can force-submit the input? This would be handy for having the "enter" key be another option to submit your inputs.

its ask and wait which means it should wait, and force submitting the input without the users permission could be annoying due to it having the off chance to force submit a unfinished sentence word, or anything that was typed, i say this because it could be multi-line. if possible it should just be "shift-enter" to submit, or like in the console set it to "enter" and to add a new line set it to "shift-enter"

SaFireGalaxy commented:
That's a good suggestion. Maybe also, Shift enter could create a new line and enter could submit it. It will act as the submit button, but not a replacement of it.

SharkPool-SP commented:
> That's a good suggestion. Maybe also, Shift enter could create a new line and enter could submit it. It will act as the submit button, but not a replacement of it.

Hi, so I read the past couple comments. I’ll consider the shift enter thing. I’ll either add it or add a block that allows/doesn’t allow enter to submit.

SaFireGalaxy commented:
Thanks so much! This beats the ask and wait block in every possible way now, I won't have to use that block anymore.

SharkPool-SP commented:
> Thanks so much! This beats the ask and wait block in every possible way now, I won't have to use that block anymore.

Thanks :)

SharkPool-SP commented:
> > > > > > Adds various blocks for custom CSS inputs. Focused on remodeling the ask () block to be better. <img alt="Screen Shot 2023-07-30 at 11 49 39 PM" width="521" src="https://camo.githubusercontent.com/a479a9b812fd016825b9651cfb1690a144b2713d2af251e603950ce9e0a4bcc1/68747470733a2f2f75736572636f6e74656e742e676974687562666173742e636f6d2f757365722d696d616765732f3133393039373337382f3235373133393230372d62353738373165652d303831342d343831372d613932372d6135653464323235323031382e706e67">
> > > > > 
> > > > > 
> > > > > Should UserInput return null instead of 'canceled' without input?
> > > > 
> > > > 
> > > > Doesn’t really matter. If there’s no response it will return empty; just like the current ask and wait block
> > > 
> > > 
> > > It does matter, actually. For parity with the actual ask block, you should return an empty string.
> > 
> > 
> > No that’s what I mean, it will return an empty string when there’s no input
> 
> It should also do that when it's cancelled. Make a boolean that is toggled when an input is cancelled or not.

No need for a boolean

SharkPool-SP commented:
> Thanks so much! This beats the ask and wait block in every possible way now, I won't have to use that block anymore.

Updated! Added that block and some more 👀

GarboMuffin commented on extensions/SharkPool/BetterInput.js:
```diff
@@ -0,0 +1,1305 @@
+            text: "box limit",
+          },
+        ],
+        menus: {
```
all of the menus should be acceptReporters: true unless they have a good reason not to

GarboMuffin commented on extensions/SharkPool/BetterInput.js:
```diff
@@ -0,0 +1,1305 @@
+        ],
+        menus: {
+          alignmentMenu: ["left", "right", "center"],
+          fontMenu: [
```
A vast majority of these fonts will not work for a *lot* of our users (anyone on ChromeOS, macOS, Linux)

Do what animated text does - stick to the standard scratch fonts. When custom fonts are merge then you can use those too. As long as it's acceptReporters: true, people will be able to manually drop in whatever they want anyways

SharkPool-SP replied:
I can remove the custom fonts stuff. Ill also try to have accept reporters in the next update whenever it comes. If you are merging this file right now, are you modifying that stuff or am I?

SharkPool-SP replied:
Saving this for last...

GarboMuffin replied:
You can do custom fonts now. See animated text

GarboMuffin commented on extensions/SharkPool/BetterInput.js:
```diff
@@ -0,0 +1,1305 @@
+      this.textBoxX = args.X;
+      this.textBoxY = args.Y * -1;
+
+      const overlays = document.querySelectorAll(".ask-box");
```
querySelectorAll is slow and prone to collisions. it would be best if you could track those elements yourself

SharkPool-SP replied:
Fixed

GarboMuffin replied:
You still use querySelectorAll in removeAskBoxes

GarboMuffin commented on extensions/SharkPool/BetterInput.js:
```diff
@@ -0,0 +1,1305 @@
+    askAndWaitForInput(args) {
+      if (this.askBoxCount < this.maxBoxCount) {
+        return new Promise((resolve) => {
+          this.askAndWait(args).then(() => {
```
pretty sure the whole new Promise thing isn't required since the thing you're wrapping is, itself, a promise

SharkPool-SP replied:
I need it to be that way so it can properly handle the various inputs. That including the ask and wait reporter block... If Im wrong, what can I do?

GarboMuffin replied:
This was not resolved

GarboMuffin commented on extensions/SharkPool/BetterInput.js:
```diff
@@ -0,0 +1,1305 @@
+          overlay.style.transition = "opacity 0.3s";
+        }
+
+        requestAnimationFrame(step);
```
are you able to do this with CSS transitions and animations instead?

SharkPool-SP replied:
Yeah sorta. Not exactly exit effects... I was planning this for a later update though since its complex to make for exiting. In my opinion its best to leave this as is until the next update rather than rushing and pushing it out... right?

SharkPool-SP replied:
Screw it, Im just gonna do it now.

GarboMuffin commented on extensions/SharkPool/BetterInput.js:
```diff
@@ -0,0 +1,1305 @@
+    }
+
+    askAndWait(args) {
+      if (this.askBoxCount < this.maxBoxCount) {
```
the normal Scratch ask blocks will wait for the old prompt to disappear, then show the new one. would it make sense to do that here too?

SharkPool-SP replied:
No need for it with the new block in this update. Since I want to allow people to have more than 1 ask box, I just added a <is waiting for input?> boolean instead. If users really wanted to wait for the old ask input to disappear before asking for the new one, they can make it themselves with the ask block. Its better this way since this extension is supposed to be a fully customizable expansion + remodel of the Ask block.

GarboMuffin commented on extensions/SharkPool/BetterInput.js:
```diff
@@ -0,0 +1,1305 @@
+          overlay.style.fontFamily = this.fontFamily;
+
+          const overlayImageContainer = document.createElement("div");
+          overlayImageContainer.style.background = `url(${this.overlayImage})`;
```
Scratch.canFetch, wrap the URL with quotes and pass it into encodeURI()

GarboMuffin replied:
This was not resolved. You still do not use Scratch.canFetch()

GarboMuffin reviewed (changes requested):
i left some comments

SharkPool-SP commented:
> Thanks so much! This beats the ask and wait block in every possible way now, I won't have to use that block anymore.

Thanks!!! Updates coming soon :)))

SharkPool-SP commented:
Fixed all issues and added new stuff!
<img width="477" alt="Screen Shot 2023-08-18 at 11 56 56 PM" src="https://github.com/TurboWarp/extensions/assets/139097378/34d812bf-8c66-4f9a-bc7e-379bf893ea64">

SharkPool-SP commented:
**3.0 is RELEASED!!!**

In this Update:
-Bug Fixes
-Infinite Buttons
-Image Setting for More Elements
-**ALL** Elements update live
-Vertical Sliders
-Updated Palette

SharkPool-SP commented:
Small Update:
![image](https://github.com/TurboWarp/extensions/assets/139097378/035e1011-d884-48d0-9013-3bec678404d1)

SharkPool-SP commented:
![image](https://github.com/TurboWarp/extensions/assets/139097378/a77fb911-06bb-4bed-83f8-daddb028ceca)
Another Update

horizontal-shading commented:
That #1029 issue wasn't a real issue. It was me complaining...

Thebloxers998 commented:
How do I change the background colour in the extension ask ui

Thebloxers998 commented:
I have a suggestion, you should make it so we can have more than one input typing thing

SharkPool-SP commented:
> I have a suggestion, you should make it so we can have more than one input typing thing

With the way I built the extension, no. As a alternative just make multiple boxes

SharkPool-SP commented:
marking as draft as im rewriting this from scratch

SharkPool-SP commented:
https://github.com/TurboWarp/extensions/pull/1940
